flink/v1.16/flink/src/main/java/org/apache/iceberg/flink/sink/shuffle/DataStatisticsOperator.java

```diff
@@ -0,0 +1,134 @@
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.flink.sink.shuffle;
+
+import org.apache.flink.api.common.state.ListState;
+import org.apache.flink.api.common.state.ListStateDescriptor;
+import org.apache.flink.api.common.typeinfo.TypeHint;
+import org.apache.flink.api.common.typeinfo.TypeInformation;
+import org.apache.flink.api.java.functions.KeySelector;
+import org.apache.flink.runtime.operators.coordination.OperatorEvent;
+import org.apache.flink.runtime.operators.coordination.OperatorEventGateway;
+import org.apache.flink.runtime.operators.coordination.OperatorEventHandler;
+import org.apache.flink.runtime.state.StateInitializationContext;
+import org.apache.flink.runtime.state.StateSnapshotContext;
+import org.apache.flink.streaming.api.operators.AbstractStreamOperator;
+import org.apache.flink.streaming.api.operators.OneInputStreamOperator;
+import org.apache.flink.streaming.runtime.streamrecord.StreamRecord;
+import org.apache.iceberg.flink.sink.shuffle.statistics.DataStatistics;
+import org.apache.iceberg.flink.sink.shuffle.statistics.DataStatisticsFactory;
+import org.apache.iceberg.relocated.com.google.common.annotations.VisibleForTesting;
+
+/**
+ * DataStatisticsOperator can help to improve data clustering based on the key.
+ *
+ * <p>DataStatisticsOperator collects traffic distribution statistics. A custom partitioner shall be
+ * attached to the DataStatisticsOperator output. The custom partitioner leverages the statistics to
+ * shuffle record to improve data clustering while maintaining relative balanced traffic
+ * distribution to downstream subtasks.
+ */
+class DataStatisticsOperator<T, K> extends AbstractStreamOperator<DataStatisticsOrRecord<T, K>>
+    implements OneInputStreamOperator<T, DataStatisticsOrRecord<T, K>>, OperatorEventHandler {
+  private static final long serialVersionUID = 1L;
+
+  // keySelector will be used to generate key from data for collecting data statistics
+  private final KeySelector<T, K> keySelector;
+  private final OperatorEventGateway operatorEventGateway;
+  private final DataStatisticsFactory<K> statisticsFactory;
+  private DataStatistics<K> localStatistics;
+  private DataStatistics<K> globalStatistics;
+  private ListState<DataStatistics<K>> globalStatisticsState;
+
+  public DataStatisticsOperator(
+      KeySelector<T, K> keySelector,
+      OperatorEventGateway operatorEventGateway,
+      DataStatisticsFactory<K> statisticsFactory) {
+    this.keySelector = keySelector;
+    this.operatorEventGateway = operatorEventGateway;
+    this.statisticsFactory = statisticsFactory;
+  }
+
+  @Override
+  public void initializeState(StateInitializationContext context) throws Exception {
+    localStatistics = statisticsFactory.createDataStatistics();
+    globalStatisticsState =
+        context
+            .getOperatorStateStore()
+            .getUnionListState(
```

Comment:
Hi @stevenzwu  as we discussed offline, I update the state from ListState to  UnionListState since ListState uses round robin pattern during rescaling and UnionlistState uses broadcast pattern which can satisfy the DataStatisticsOperator need better(everyone should have the same global data statistics)